File: Python/Graphs/FloydWarshall.py
"""
Floyd-Warshall Algorithm Implementation

Time Complexity: O(V^3)
Space Complexity: O(V^2)
"""

def floyd_warshall(graph):
    n = len(graph)
    # Initialize distances. Use float('inf') for no edge, and 0 for i == j.
    dist = []
    for i in range(n):
        row = []
        for j in range(n):
            if i == j:
                row.append(0)
            elif graph[i][j] == 0:
                row.append(float('inf'))
            else:
                row.append(graph[i][j])
        dist.append(row)

    for k in range(n):
        for i in range(n):
            for j in range(n):
                if dist[i][k] != float('inf') and dist[k][j] != float('inf') and \
                   dist[i][k] + dist[k][j] < dist[i][j]:
                    dist[i][j] = dist[i][k] + dist[k][j]

    return dist

# Example usage:
# if __name__ == "__main__":
#     graph = [
#         [0, 5, 0, 10],
#         [0, 0, 3, 0],
#         [0, 0, 0, 1],
#         [0, 0, 0, 0]
#     ]
#     print(floyd_warshall(graph))

File: CPP/Graphs/FloydWarshall.cpp
/**
 * Floyd-Warshall Algorithm Implementation
 * 
 * Time Complexity: O(V^3)
 * Space Complexity: O(V^2)
 */

#include <iostream>
#include <vector>

#define INF 99999

void floydWarshall(std::vector<std::vector<int>>& graph) {
    int V = graph.size();
    std::vector<std::vector<int>> dist = graph;

    for (int i = 0; i < V; i++) {
        for (int j = 0; j < V; j++) {
            if (i != j && dist[i][j] == 0) dist[i][j] = INF;
        }
    }

    for (int k = 0; k < V; k++) {
        for (int i = 0; i < V; i++) {
            for (int j = 0; j < V; j++) {
                if (dist[i][k] + dist[k][j] < dist[i][j])
                    dist[i][j] = dist[i][k] + dist[k][j];
            }
        }
    }

    for (int i = 0; i < V; i++) {
        for (int j = 0; j < V; j++) {
            if (dist[i][j] == INF) std::cout << "INF ";
            else std::cout << dist[i][j] << "   ";
        }
        std::cout << std::endl;
    }
}

/*
int main() {
    std::vector<std::vector<int>> graph = {
        {0, 5, INF, 10},
        {INF, 0, 3, INF},
        {INF, INF, 0, 1},
        {INF, INF, INF, 0}
    };
    floydWarshall(graph);
    return 0;
}
*/
